### PR TITLE
Core: Add ManifestWrite benchmark

### PR DESCRIPTION
**Pull request by Fokko**

### CODE REVIEW

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void writeManifestFile() throws IOException {
+    baseDir = Files.createTempDir().getAbsolutePath();
+    manifestListFile = String.format("%s/%s.avro", baseDir, UUID.randomUUID());
+
+    try (ManifestListWriter listWriter =
+        ManifestLists.write(1, org.apache.iceberg.Files.localOutput(manifestListFile), 0, 1L, 0)) {
```
Comment (dramaticlly) [inline core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:92]: shall we also add a parameter for format version? I guess if would be great if we can test both format version but not really needed in same PR

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+ *
+ * <p>To run this benchmark: <code>
+ *   ./gradlew :iceberg-core:jmh -PjmhIncludeRegex=ManifestWriteBenchmark
+ * </code>
+ */
+@Fork(1)
+@State(Scope.Benchmark)
+@Measurement(iterations = 5)
+@BenchmarkMode(Mode.SingleShotTime)
+@Timeout(time = 1000, timeUnit = TimeUnit.HOURS)
```
Comment (dramaticlly) [inline core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:56]: Questions: do we need 1000 hours as benchmark timeout? This seems suspicious, I ran the read benchmark locally I think 1000 seconds shall be more than enough.

**Review by dramaticlly — APPROVED** [review]

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 2 above)
Comment (aokolnychyi) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:56]: I agree, the timeout seems off.

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 1 above)
Comment (aokolnychyi) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:92]: I agree, it would be nice to cover both formats. We can use `@Param` annotation.

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+      baseDir = null;
+    }
+
+    manifestListFile = null;
+  }
+
+  @Benchmark
+  @Threads(1)
+  public void writeManifestFile() throws IOException {
+    baseDir = Files.createTempDir().getAbsolutePath();
```
Comment (aokolnychyi) [inline core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:88]: Minor: We frequently use `this.varName` prefix when setting (not getting) instance variable to highlight that it is an instance variable, not a local variable. It is optional, up to you @Fokko. If you decide to apply that, must be done to all places that assign instance variables.

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+        OutputFile manifestFile =
+            org.apache.iceberg.Files.localOutput(
+                String.format("%s/%s.avro", baseDir, UUID.randomUUID()));
+
+        ManifestWriter<DataFile> writer =
+            ManifestFiles.write(1, PartitionSpec.unpartitioned(), manifestFile, 1L);
+        try (ManifestWriter<DataFile> finalWriter = writer) {
+          for (int j = 0; j < NUM_ROWS; j++) {
+            DataFile dataFile =
+                DataFiles.builder(PartitionSpec.unpartitioned())
```
Comment (aokolnychyi) [inline core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:103]: It would be nice to cover partitioned and delete manifests (could be a separate PR).

**Review by aokolnychyi — COMMENTED** [review]
This seems really close. I agree with comments from @dramaticlly. Could you take a look, @Fokko? Let's get this in.

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 2 above)
Comment (Fokko) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:56]: I left this out since this was not part of the read benchmark, but added it 👍

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 3 above)
Comment (Fokko) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:88]: As a pythonista; explicit is better than implicit. I've added `this.`

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 4 above)
Comment (Fokko) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:103]: Let's do it in a separate PR 👍

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java (hunk 2 above)
Comment (Fokko) [reply core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:57]:
```suggestion
@Timeout(time = 5, timeUnit = TimeUnit.MINUTES)
```

File: core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+      baseDir = null;
+    }
+
+    manifestListFile = null;
+  }
+
+  @State(Scope.Benchmark)
+  public static class BenchmarkState {
+    @Param({"1", "2"})
+    public int formatVersion;
```
Comment (dramaticlly) [inline core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:89]:
```suggestion
    int formatVersion;
```

checkstyle does not like it and I believe it does not need to be public if we always access it via state

**Comment by aokolnychyi** [comment]
There was a style violation:

```
Error: eckstyle] [ERROR] /home/runner/work/iceberg/iceberg/core/src/jmh/java/org/apache/iceberg/ManifestWriteBenchmark.java:89:9: Variable 'formatVersion' must be private and have accessor methods. [VisibilityModifier]
```

**Review by aokolnychyi — APPROVED** [review]
